### DISCUSSION TITLE
Python project setup being misclassified as HaProxy when it should be INI

### DISCUSSION THREAD

**Discussion by RivenSkaye** [github-linguist/linguist#6489] [Classification] [answered]
_all hail the template_

## Problem description
I'm seeing not just my own, but big-name Python projects that include a `setup.cfg` for configuration of tools such as flake8, mypy and other dev support utilities as well as installer configuration that happens in the same file getting the file flagged as `HaProxy` instead of `INI`. PR #4303 mentions haproxy configurations being allowed to have any arbitrary name, but in the first several pages of [searching for setup.cfg flagged as haproxy](https://github.com/search?q=path%3A*.cfg+path%3A**%2Fsetup.cfg++language%3AHAProxy&type=code&ref=advsearch&l=HAProxy) I can't find a single file that seems to _actually_ be for haproxy. This also means that setting `**/*.cfg linguist-detectable` in `.gitattributes` will also provide incorrect results on the repository stats.

As a result, changing the `lang:` search parameter shows a lot of example files that either have shebangs (yet aren't valid files, triggering an error when accidentally run) or that use modeline capabilities to override linguist.
There seem to be _some_ instances of [setup.cfg being considered `INI` files](https://github.com/search?q=path%3A*.cfg+path%3A**%2Fsetup.cfg++language%3AINI&type=code&ref=advsearch) when searching for it specifically, but most of those seem to be very short.

### URL of the affected repository:
[The expansive search](https://github.com/search?q=path%3A*.cfg+path%3A**%2Fsetup.cfg++language%3AHAProxy&type=code&ref=advsearch&l=HAProxy) provides a better indicator since it flags languages per-file, which is the issue here.

If anyone happens to know a repository providing only or mostly example Python setups, or one that includes the files and has them mistagged on the stats, please comment a link and I'll edit it in.

### Expected language:
INI

### Detected language:
HaProxy

**Comment by lildude** [comment, marked as answer]
Almost all of these are going to be because of the `.cfg` extension matching both HAProxy and INI and Linguist falling through to the classifier as there isn't a heuristic for the `.cfg` extension. The classifier is trained on the samples provided in Linguist and in the case of the `.cfg` extension, there are 4 samples for HAProxy and only tiny, and pretty poor 5 year old sample for INI.

There are a few things that can be done to improve this:

1. Add a heuristic (it's a regex) for `.cfg` files to differentiate between HAProxy and INI
2. Add more INI representative `.cfg` samples to better train the classifier

I'm loath to add `setup.cfg` as a filename for INI as it's a very generic filename so will result in all files of that name being identified as INI, even those with shebangs as the shebang strategy comes after the filename stategy.

If you've go the time, feel free to submit a PR for the above.
